### DISCUSSION TITLE
Settings in pyrightconfig.json seemingly ignored

### DISCUSSION THREAD

**Discussion by Telofy** [microsoft/pyright#7323] [Q&A] [answered]
I'm trying to set up my VS Code such that it alerts me of my own type errors (strict mode) but not of those in third-party libraries. My code is in `/home/telofy/projects/`, the third-party libraries are in `/home/telofy/miniconda3/`.

One attempt:

```
{
    "typeCheckingMode": "strict",
    "ignore": [
        "/home/telofy/miniconda3/"
    ]
}
```

The `ignore` has no effect (?). When I call CCXT's `fetch_balance()` (a third-party library method), I still get `Type of "fetch_balance" is partially unknown`. What's good is that strict mode is still applied to my own code.

Another attempt:

```
{
    "typeCheckingMode": "standard",
    "strict": [
        "/home/telofy/projects/"
    ]
}
```

The `strict` has no effect (?). Now all my code and that of third-party libraries is checked with `standard` mode and I get virtually no type errors. I also tried `/home/telofy/projects/**` and `/home/telofy/projects/**/*`.

The “throw everything at it” approach doesn't work either:

```
{
    "typeCheckingMode": "strict",
    "root": "/home/telofy/projects/",
    "strict": [
        "/home/telofy/projects/"
    ],
    "ignore": [
        "/home/telofy/miniconda3/"
    ],
    "exclude": [
        "/home/telofy/miniconda3/"
    ],
    "executionEnvironments": [
        {
            "root": "/home/telofy/miniconda3/",
            "pythonVersion": "3.0"
        }
    ]
}
```

(I hoped that setting a Python version from before type annotations were introduced might help.)

There are no type-checking-relevant settings in pyproject.toml or my workspace configuration.

**Comment by erictraut** [comment, marked as answer]
It sounds like you have multiple Python projects in your "projects" folder. I presume that each project has its own subdirectory. When you open a project in VS Code, are you opening the "projects" folder, or are you opening a subdirectory corresponding to a project within the "projects" folder? Typically, you'd do the latter. Each project would have its own pyrightconfig.json or pyproject.toml file. It's fine for multiple projects to share a single Python environment.

The pyrightconfig.json config file should not contain any absolute paths like you have shown in your example. This config file should be checked in to your source repo and shared among users, and absolute paths break this.

For a given project, pyright by default considers all ".py" and ".pyi" files as part of the project and will apply type checking to those files. It auto-excludes a virtual environment subdirectory if it detects one within your project. If you would like to exclude other files, you can add an "exclude" list in your config file.

Of the files that pyright considers part of the project, you can choose to "ignore" some of them. This simply suppresses all of the diagnostics produced for those source files. I don't think that's what you want here.

If you enable strict mode for a file, pyright will report any use of symbols whose types are not fully known (i.e. they are "unknown" or "partially unknown"). That's why you're seeing errors like "Type of "fetch_balance" is partially unknown". If you are importing from libraries that do not contain full type information, you will not be able to use strict mode without seeing these errors. In this case, you either need to supply your own type information (in the form of user-maintained type stubs for those libraries) or disable strict mode.

**Reply by Telofy** [reply]
Thank you for the explanation!

I've changed the paths to relative ones, but that didn't cause a problem since we would (in theory) use pyproject.toml for shared settings, so that I went with pyrightconfig.json for my personal overrides (and put it in .gitignore).

> If you enable strict mode for a file, pyright will report any use of symbols whose types are not fully known (i.e. they are "unknown" or "partially unknown"). That's why you're seeing errors like "Type of "fetch_balance" is partially unknown". If you are importing from libraries that do not contain full type information, you will not be able to use strict mode without seeing these errors. In this case, you either need to supply your own type information (in the form of user-maintained type stubs for those libraries) or disable strict mode.

Aw, that's unfortunate. I had hoped that I had found a way to change that. I've tried creating stub files, but it was a lot of work. Often I couldn't just selectively type one method in the stub file but needed to be fairly comprehensive to make it work. And what I did was just to add `Any` wherever a type was missing, so it wasn't even any use.

I'd love to have a setting that causes Pyright to assume `Unknown` = `Any` for all the code in certain directories (e.g., **/site-packages).

**Reply by erictraut** [reply]
You can create a minimal type stub for a given library that tells a type checker that all symbols imported from that library should be treated as `Any`. This leverages the functionality in [PEP 562](https://peps.python.org/pep-0562/). To do this, create a project-local "typings" directory. Then create a subdirectory with the name of your library (its top-level module name). Within that subdirectory, create an `__init__.pyi` file that contains the following:

```python
def __getattr__(name: str) -> Any:
    ...
```

You can then optionally add select class and function types to the type stub to augment it as you see fit.

Note that this largely defeats the purpose of full strict-mode type checking because these `Any` types tend to propagate through the code during type analysis, so the type checker is blinded to many issues. If you don't have full types for the libraries you're using, I recommend against using full strict mode. You can enable a subset of the strict-mode rules as you see fit for your project.

**Reply by Telofy** [reply]
Oh, exciting! I hadn't tried the `__getattr__` trick! I have gotten it to work in a simplified case. One caveat seems to be that each directory of the third-party library requires its own `__init__.pyi` override.

E.g., this works:

![image](https://github.com/microsoft/pyright/assets/223663/8a5744de-0b6e-4706-88d0-b9d77234f71f)

But this does not:

![image](https://github.com/microsoft/pyright/assets/223663/9117f962-7017-4403-9ce0-b69eea1ce370)

Our actual code uses many different parts of the library, so I won't get around rebuilding most of the directory tree of the library, but this is promising! (Does it obscure whatever partial-but-useful typing the library does have?)

It is somewhat purpose-defeating of course, but realistically the alternatives are that we'll use standard mode for all code and so also lose out on many useful rules in our own code, that we'll litter the code with `# pyright: ignore`, or that we'll try to mentally block out the red squiggles. :-/

Thanks so much for your help!

(I still hope there'll be a config parameter for this at some point. ^.^)
